File: ios/AlpineApp/AlpineApp/Services/AlpineRpcService.swift
import Foundation

final class AlpineRpcService: Sendable {
    private let client: JsonRpcClient

    init(client: JsonRpcClient) {
        self.client = client
    }

    // MARK: - Status

    func getStatus() async throws -> ServerStatus {
        try await client.call(method: "getStatus")
    }

    // MARK: - Queries

    func startQuery(
        queryString: String,
        groupName: String = "",
        autoHaltLimit: Int64 = 100,
        peerDescMax: Int64 = 50
    ) async throws -> Int64 {
        let response: QueryResponse = try await client.call(
            method: "startQuery",
            params: [
                "queryString": queryString,
                "groupName": groupName,
                "autoHaltLimit": autoHaltLimit,
                "peerDescMax": peerDescMax
            ]
        )
        return response.queryId
    }

    func queryInProgress(queryId: Int64) async throws -> Bool {
        try await client.call(method: "queryInProgress", params: ["queryId": queryId])
    }

    func getQueryStatus(queryId: Int64) async throws -> QueryStatusResponse {
        try await client.call(method: "getQueryStatus", params: ["queryId": queryId])
    }

    func getQueryResults(queryId: Int64) async throws -> QueryResultsResponse {
        try await client.call(method: "getQueryResults", params: ["queryId": queryId])
    }

    func pauseQuery(queryId: Int64) async throws -> Bool {
        try await client.call(method: "pauseQuery", params: ["queryId": queryId])
    }

    func resumeQuery(queryId: Int64) async throws -> Bool {
        try await client.call(method: "resumeQuery", params: ["queryId": queryId])
    }

    func cancelQuery(queryId: Int64) async throws -> Bool {
        try await client.call(method: "cancelQuery", params: ["queryId": queryId])
    }

    // MARK: - Peers

    func getAllPeers() async throws -> [Int64] {
        try await client.call(method: "getAllPeers")
    }

    func getPeer(peerId: Int64) async throws -> PeerDetail {
        try await client.call(method: "getPeer", params: ["peerId": peerId])
    }

    func addPeer(ipAddress: String, port: Int64) async throws -> Bool {
        try await client.call(method: "addPeer", params: ["ipAddress": ipAddress, "port": port])
    }

    func getPeerId(ipAddress: String, port: Int64) async throws -> Int64 {
        try await client.call(method: "getPeerId", params: ["ipAddress": ipAddress, "port": port])
    }

    func activatePeer(peerId: Int64) async throws -> Bool {
        try await client.call(method: "activatePeer", params: ["peerId": peerId])
    }

    func deactivatePeer(peerId: Int64) async throws -> Bool {
        try await client.call(method: "deactivatePeer", params: ["peerId": peerId])
    }

    func pingPeer(peerId: Int64) async throws -> Bool {
        try await client.call(method: "pingPeer", params: ["peerId": peerId])
    }

    // MARK: - Filters

    func excludeHost(ipAddress: String) async throws -> Bool {
        try await client.call(method: "excludeHost", params: ["ipAddress": ipAddress])
    }

    func excludeSubnet(subnetIpAddress: String, subnetMask: String) async throws -> Bool {
        try await client.call(
            method: "excludeSubnet",
            params: ["subnetIpAddress": subnetIpAddress, "subnetMask": subnetMask]
        )
    }

    func allowHost(ipAddress: String) async throws -> Bool {
        try await client.call(method: "allowHost", params: ["ipAddress": ipAddress])
    }

    func allowSubnet(subnetIpAddress: String, subnetMask: String) async throws -> Bool {
        try await client.call(
            method: "allowSubnet",
            params: ["subnetIpAddress": subnetIpAddress, "subnetMask": subnetMask]
        )
    }

    func listExcludedHosts() async throws -> [String] {
        try await client.call(method: "listExcludedHosts")
    }

    func listExcludedSubnets() async throws -> [Subnet] {
        try await client.call(method: "listExcludedSubnets")
    }

    // MARK: - Groups

    func createGroup(name: String, description: String = "") async throws -> Int64 {
        try await client.call(
            method: "createGroup",
            params: ["name": name, "description": description]
        )
    }

    func deleteGroup(groupId: Int64) async throws -> Bool {
        try await client.call(method: "deleteGroup", params: ["groupId": groupId])
    }

    func listGroups() async throws -> [Int64] {
        try await client.call(method: "listGroups")
    }

    func getGroupInfo(groupId: Int64) async throws -> GroupInfo {
        try await client.call(method: "getGroupInfo", params: ["groupId": groupId])
    }

    func getDefaultGroupInfo() async throws -> GroupInfo {
        try await client.call(method: "getDefaultGroupInfo")
    }

    func getGroupPeerList(groupId: Int64) async throws -> [Int64] {
        try await client.call(method: "getGroupPeerList", params: ["groupId": groupId])
    }

    func addPeerToGroup(groupId: Int64, peerId: Int64) async throws -> Bool {
        try await client.call(
            method: "addPeerToGroup",
            params: ["groupId": groupId, "peerId": peerId]
        )
    }

    func removePeerFromGroup(groupId: Int64, peerId: Int64) async throws -> Bool {
        try await client.call(
            method: "removePeerFromGroup",
            params: ["groupId": groupId, "peerId": peerId]
        )
    }

    // MARK: - Modules

    func registerModule(libraryPath: String, bootstrapSymbol: String) async throws -> Int64 {
        try await client.call(
            method: "registerModule",
            params: ["libraryPath": libraryPath, "bootstrapSymbol": bootstrapSymbol]
        )
    }

    func unregisterModule(moduleId: Int64) async throws -> Bool {
        try await client.call(method: "unregisterModule", params: ["moduleId": moduleId])
    }

    func loadModule(moduleId: Int64) async throws -> Bool {
        try await client.call(method: "loadModule", params: ["moduleId": moduleId])
    }

    func unloadModule(moduleId: Int64) async throws -> Bool {
        try await client.call(method: "unloadModule", params: ["moduleId": moduleId])
    }

    func listActiveModules() async throws -> [Int64] {
        try await client.call(method: "listActiveModules")
    }

    func listAllModules() async throws -> [Int64] {
        try await client.call(method: "listAllModules")
    }

    func getModuleInfo(moduleId: Int64) async throws -> ModuleInfo {
        try await client.call(method: "getModuleInfo", params: ["moduleId": moduleId])
    }

    // MARK: - Lifecycle

    func shutdown() {
        Task { await client.shutdown() }
    }
}

File: android/AlpineApp/app/src/main/java/com/alpine/app/data/rpc/AlpineRpcService.kt
package com.alpine.app.data.rpc

import com.alpine.app.data.model.FilterModels
import com.alpine.app.data.model.GroupInfo
import com.alpine.app.data.model.ModuleInfo
import com.alpine.app.data.model.PeerDetail
import com.alpine.app.data.model.PeerResources
import com.alpine.app.data.model.QueryResultsResponse
import com.alpine.app.data.model.QueryStatusResponse
import com.alpine.app.data.model.ResourceDesc
import com.alpine.app.data.model.ServerStatus
import com.google.gson.JsonArray
import com.google.gson.JsonObject

class AlpineRpcService(
    baseUrl: String,
    tlsConfig: TlsConfig = TlsConfig(),
    host: String = ""
) {
    private val client = JsonRpcClient(baseUrl, tlsConfig, host)

    // --- Status ---

    suspend fun getStatus(): ServerStatus {
        val result = client.call("getStatus").asJsonObject
        return ServerStatus(
            status = result.get("status")?.asString ?: "unknown",
            version = result.get("version")?.asString ?: "unknown"
        )
    }

    // --- Query ---

    suspend fun startQuery(
        queryString: String,
        groupName: String = "",
        autoHaltLimit: Long = 100,
        peerDescMax: Long = 50
    ): Long {
        val params = JsonObject().apply {
            addProperty("queryString", queryString)
            if (groupName.isNotBlank()) addProperty("groupName", groupName)
            addProperty("autoHaltLimit", autoHaltLimit)
            addProperty("peerDescMax", peerDescMax)
        }
        val result = client.call("startQuery", params).asJsonObject
        return result.get("queryId").asLong
    }

    suspend fun queryInProgress(queryId: Long): Boolean {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("queryInProgress", params).asJsonObject
        return result.get("inProgress").asBoolean
    }

    suspend fun getQueryStatus(queryId: Long): QueryStatusResponse {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("getQueryStatus", params).asJsonObject
        return QueryStatusResponse(
            inProgress = true,
            totalPeers = result.get("totalPeers")?.asLong ?: 0,
            peersQueried = result.get("peersQueried")?.asLong ?: 0,
            numPeerResponses = result.get("numPeerResponses")?.asLong ?: 0,
            totalHits = result.get("totalHits")?.asLong ?: 0
        )
    }

    suspend fun getQueryResults(queryId: Long): QueryResultsResponse {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("getQueryResults", params).asJsonObject
        val peers = result.getAsJsonArray("peers")?.map { peerEl ->
            val peer = peerEl.asJsonObject
            val resources = peer.getAsJsonArray("resources")?.map { resEl ->
                val res = resEl.asJsonObject
                val locators = res.getAsJsonArray("locators")?.map { it.asString } ?: emptyList()
                ResourceDesc(
                    resourceId = res.get("resourceId")?.asLong ?: 0,
                    size = res.get("size")?.asLong ?: 0,
                    description = res.get("description")?.asString ?: "",
                    locators = locators
                )
            } ?: emptyList()
            PeerResources(
                peerId = peer.get("peerId")?.asLong ?: 0,
                resources = resources
            )
        } ?: emptyList()
        return QueryResultsResponse(peers)
    }

    suspend fun pauseQuery(queryId: Long): Boolean {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("pauseQuery", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun resumeQuery(queryId: Long): Boolean {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("resumeQuery", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun cancelQuery(queryId: Long): Boolean {
        val params = JsonObject().apply { addProperty("queryId", queryId) }
        val result = client.call("cancelQuery", params).asJsonObject
        return result.get("cancelled")?.asBoolean ?: false
    }

    // --- Peers ---

    suspend fun getAllPeers(): List<Long> {
        val result = client.call("getAllPeers").asJsonObject
        return result.getAsJsonArray("peerIds")?.map { it.asLong } ?: emptyList()
    }

    suspend fun getPeer(peerId: Long): PeerDetail {
        val params = JsonObject().apply { addProperty("peerId", peerId) }
        val result = client.call("getPeer", params).asJsonObject
        return PeerDetail(
            peerId = result.get("peerId")?.asLong ?: peerId,
            ipAddress = result.get("ipAddress")?.asString ?: "",
            port = result.get("port")?.asLong ?: 0,
            lastRecvTime = result.get("lastRecvTime")?.asLong ?: 0,
            lastSendTime = result.get("lastSendTime")?.asLong ?: 0,
            avgBandwidth = result.get("avgBandwidth")?.asLong ?: 0,
            peakBandwidth = result.get("peakBandwidth")?.asLong ?: 0
        )
    }

    suspend fun addPeer(ipAddress: String, port: Long): Boolean {
        val params = JsonObject().apply {
            addProperty("ipAddress", ipAddress)
            addProperty("port", port)
        }
        val result = client.call("addPeer", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun getPeerId(ipAddress: String, port: Long): Long {
        val params = JsonObject().apply {
            addProperty("ipAddress", ipAddress)
            addProperty("port", port)
        }
        val result = client.call("getPeerId", params).asJsonObject
        return result.get("peerId")?.asLong ?: 0
    }

    suspend fun activatePeer(peerId: Long): Boolean {
        val params = JsonObject().apply { addProperty("peerId", peerId) }
        val result = client.call("activatePeer", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun deactivatePeer(peerId: Long): Boolean {
        val params = JsonObject().apply { addProperty("peerId", peerId) }
        val result = client.call("deactivatePeer", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun pingPeer(peerId: Long): Boolean {
        val params = JsonObject().apply { addProperty("peerId", peerId) }
        val result = client.call("pingPeer", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    // --- Network Filters ---

    suspend fun excludeHost(ipAddress: String): Boolean {
        val params = JsonObject().apply { addProperty("ipAddress", ipAddress) }
        val result = client.call("excludeHost", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun excludeSubnet(subnetIpAddress: String, subnetMask: String): Boolean {
        val params = JsonObject().apply {
            addProperty("subnetIpAddress", subnetIpAddress)
            addProperty("subnetMask", subnetMask)
        }
        val result = client.call("excludeSubnet", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun allowHost(ipAddress: String): Boolean {
        val params = JsonObject().apply { addProperty("ipAddress", ipAddress) }
        val result = client.call("allowHost", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun allowSubnet(subnetIpAddress: String, subnetMask: String): Boolean {
        val params = JsonObject().apply {
            addProperty("subnetIpAddress", subnetIpAddress)
            addProperty("subnetMask", subnetMask)
        }
        val result = client.call("allowSubnet", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun listExcludedHosts(): List<String> {
        val result = client.call("listExcludedHosts").asJsonObject
        return result.getAsJsonArray("hosts")?.map { it.asString } ?: emptyList()
    }

    suspend fun listExcludedSubnets(): List<FilterModels.Subnet> {
        val result = client.call("listExcludedSubnets").asJsonObject
        return result.getAsJsonArray("subnets")?.map { el ->
            val obj = el.asJsonObject
            FilterModels.Subnet(
                ipAddress = obj.get("ipAddress")?.asString ?: "",
                netMask = obj.get("netMask")?.asString ?: ""
            )
        } ?: emptyList()
    }

    // --- Groups ---

    suspend fun createGroup(name: String, description: String = ""): Long {
        val params = JsonObject().apply {
            addProperty("name", name)
            if (description.isNotBlank()) addProperty("description", description)
        }
        val result = client.call("createGroup", params).asJsonObject
        return result.get("groupId").asLong
    }

    suspend fun deleteGroup(groupId: Long): Boolean {
        val params = JsonObject().apply { addProperty("groupId", groupId) }
        val result = client.call("deleteGroup", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun listGroups(): List<Long> {
        val result = client.call("listGroups").asJsonObject
        return result.getAsJsonArray("groupIds")?.map { it.asLong } ?: emptyList()
    }

    suspend fun getGroupInfo(groupId: Long): GroupInfo {
        val params = JsonObject().apply { addProperty("groupId", groupId) }
        return parseGroupInfo(client.call("getGroupInfo", params).asJsonObject)
    }

    suspend fun getDefaultGroupInfo(): GroupInfo {
        return parseGroupInfo(client.call("getDefaultGroupInfo").asJsonObject)
    }

    suspend fun getGroupPeerList(groupId: Long): List<Long> {
        val params = JsonObject().apply { addProperty("groupId", groupId) }
        val result = client.call("getGroupPeerList", params).asJsonObject
        return result.getAsJsonArray("peerIds")?.map { it.asLong } ?: emptyList()
    }

    suspend fun addPeerToGroup(groupId: Long, peerId: Long): Boolean {
        val params = JsonObject().apply {
            addProperty("groupId", groupId)
            addProperty("peerId", peerId)
        }
        val result = client.call("addPeerToGroup", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun removePeerFromGroup(groupId: Long, peerId: Long): Boolean {
        val params = JsonObject().apply {
            addProperty("groupId", groupId)
            addProperty("peerId", peerId)
        }
        val result = client.call("removePeerFromGroup", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    // --- Modules ---

    suspend fun registerModule(libraryPath: String, bootstrapSymbol: String): Long {
        val params = JsonObject().apply {
            addProperty("libraryPath", libraryPath)
            addProperty("bootstrapSymbol", bootstrapSymbol)
        }
        val result = client.call("registerModule", params).asJsonObject
        return result.get("moduleId").asLong
    }

    suspend fun unregisterModule(moduleId: Long): Boolean {
        val params = JsonObject().apply { addProperty("moduleId", moduleId) }
        val result = client.call("unregisterModule", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun loadModule(moduleId: Long): Boolean {
        val params = JsonObject().apply { addProperty("moduleId", moduleId) }
        val result = client.call("loadModule", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun unloadModule(moduleId: Long): Boolean {
        val params = JsonObject().apply { addProperty("moduleId", moduleId) }
        val result = client.call("unloadModule", params).asJsonObject
        return result.get("success")?.asBoolean ?: false
    }

    suspend fun listActiveModules(): List<Long> {
        val result = client.call("listActiveModules").asJsonObject
        return result.getAsJsonArray("moduleIds")?.map { it.asLong } ?: emptyList()
    }

    suspend fun listAllModules(): List<Long> {
        val result = client.call("listAllModules").asJsonObject
        return result.getAsJsonArray("moduleIds")?.map { it.asLong } ?: emptyList()
    }

    suspend fun getModuleInfo(moduleId: Long): ModuleInfo {
        val params = JsonObject().apply { addProperty("moduleId", moduleId) }
        val result = client.call("getModuleInfo", params).asJsonObject
        return ModuleInfo(
            moduleId = result.get("moduleId")?.asLong ?: moduleId,
            moduleName = result.get("moduleName")?.asString ?: "",
            description = result.get("description")?.asString ?: "",
            version = result.get("version")?.asString ?: "",
            libraryPath = result.get("libraryPath")?.asString ?: "",
            bootstrapSymbol = result.get("bootstrapSymbol")?.asString ?: "",
            activeTime = result.get("activeTime")?.asLong ?: 0
        )
    }

    fun shutdown() {
        client.shutdown()
    }

    private fun parseGroupInfo(obj: JsonObject): GroupInfo {
        return GroupInfo(
            groupId = obj.get("groupId")?.asLong ?: 0,
            groupName = obj.get("groupName")?.asString ?: "",
            description = obj.get("description")?.asString ?: "",
            numPeers = obj.get("numPeers")?.asLong ?: 0,
            totalQueries = obj.get("totalQueries")?.asLong ?: 0,
            totalResponses = obj.get("totalResponses")?.asLong ?: 0
        )
    }
}
